devnet-sdk/system/periphery/go-ethereum/fees.go

```diff
@@ -0,0 +1,81 @@
+package goethereum
+
+import (
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+var (
+	// Ensure that the feeEstimator implements the FeeEstimator interface
+	_ FeeEstimator = (*eip1559eeEstimator)(nil)
+)
+
+// FeeEstimator is a generic fee estimation interface (not specific to EIP-1559)
+type FeeEstimator interface {
+	EstimateFees(ctx context.Context, opts *bind.TransactOpts) (*bind.TransactOpts, error)
+}
+
+// eip1559eeEstimator is a fee estimator that uses EIP-1559 fee estimation
+type eip1559eeEstimator struct {
+	// Access to the Ethereum client is needed to get the fee information from the chain
+	client FeeEthClient
+
+	// The tip multiplier is used to increase the maxPriorityFeePerGas (GasTipCap) by a factor
+	tipMultiplier *big.Int
+}
+
+func NewEIP1559FeeEstimator(client FeeEthClient) *eip1559eeEstimator {
+	return &eip1559eeEstimator{
+		client:        client,
+		tipMultiplier: big.NewInt(1),
+	}
+}
+
+func (f *eip1559eeEstimator) WithTipMultiplier(multiplier *big.Int) *eip1559eeEstimator {
+	newF := *f
+	newF.tipMultiplier = multiplier
+
+	return &newF
+}
+
+func (f *eip1559eeEstimator) EstimateFees(ctx context.Context, opts *bind.TransactOpts) (*bind.TransactOpts, error) {
+	newOpts := *opts
+
+	// Add a gas tip cap if needed
```

Comment:
I'm wondering if that's the least surprising behavior.

IIUC, the first estimation wins, when really the last one (closest to submission) would have the best chance of being accurate, no?

Reply:
How do you mean?

This is the standard estimation way - the two steps work together as one, it's not two separate options. The `nil` check only provides a way for users to skip the estimation and provide custom values for either of the two properties that determine the transaction pricing.

Reply:
If you reuse the object then yes, this is exactly what is going to happen. The pattern for estimation with some common options is smth like:

```go
commonOpts := &bind.TransactionOpts{ ... }

tx1Opts := EstimateFees(commonOpts)
tx1, err := ...

tx2Opts := EstimateFees(commonOpts)
tx2, err := ...
```

the `commonOpts` will never be modified though (there is a test for that).